File: java-nick/src/main/java/com/p2pfs/crypto/Hkdf.java
package com.p2pfs.crypto;

import org.bouncycastle.crypto.digests.SHA256Digest;
import org.bouncycastle.crypto.generators.HKDFBytesGenerator;
import org.bouncycastle.crypto.params.HKDFParameters;

import java.nio.charset.StandardCharsets;

// HKDF-SHA256 key derivation (RFC 5869)
public final class Hkdf {

    private Hkdf() {}

    // ikm = shared secret, salt = optional nonces, info = protocol label, length = output bytes
    public static byte[] derive(byte[] ikm, byte[] salt, String info, int length) {
        HKDFBytesGenerator hkdf = new HKDFBytesGenerator(new SHA256Digest());
        hkdf.init(new HKDFParameters(ikm, salt, info.getBytes(StandardCharsets.UTF_8)));
        byte[] okm = new byte[length];
        hkdf.generateBytes(okm, 0, length);
        return okm;
    }
}

File: python-alice/crypto/hkdf.py
from typing import Optional

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.hkdf import HKDF


def derive_key(
    shared_secret: bytes,
    salt: Optional[bytes] = None,
    info: bytes = b"session key",
) -> bytes:
    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        info=info,
    )
    return hkdf.derive(shared_secret)
